test/infrastructure/docker/controllers/dockermachine_controller.go

```diff
@@ -187,24 +188,34 @@ func (r *DockerMachineReconciler) reconcileNormal(ctx context.Context, machine *
 		}
 	}
 
-	bootstrapData, err := r.getBootstrapData(ctx, machine)
-	if err != nil {
-		r.Log.Error(err, "failed to get bootstrap data")
-		return ctrl.Result{}, nil
-	}
+	// if the machine isn't bootstrapped, only then run bootstrap scripts
+	if !dockerMachine.Spec.Bootstrapped {
+		bootstrapData, err := r.getBootstrapData(ctx, machine)
+		if err != nil {
+			r.Log.Error(err, "failed to get bootstrap data")
+			return ctrl.Result{}, nil
+		}
 
-	timeoutctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
-	defer cancel()
-	// Run the bootstrap script. Simulates cloud-init.
-	if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData); err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
+		timeoutctx, cancel := context.WithTimeout(ctx, 3*time.Minute)
+		defer cancel()
+		// Run the bootstrap script. Simulates cloud-init.
+		if err := externalMachine.ExecBootstrap(timeoutctx, bootstrapData); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to exec DockerMachine bootstrap")
+		}
+
+		dockerMachine.Spec.Bootstrapped = true
 	}
 
 	// Usually a cloud provider will do this, but there is no docker-cloud provider.
-	if err := externalMachine.SetNodeProviderID(ctx); err != nil {
+	// Retry to account for momentary changes in load balancer state
+	if err := wait.ExponentialBackoff(newBackoff(), func() (bool, error) {
```

Comment:
let's let the reconciler do the retries. if this fails and since it's now reentrant this retry is not necessary

Reply:
fair enough